### ISSUE TITLE
libtorch gives different results on each run on certain machines

### ISSUE THREAD

**Issue by jamesrobertwilliams** [pytorch/pytorch#76302]
### 🐛 Describe the bug

I traced a model and I am using it within libtorch, but for some bizzare reason, I notice that the results are different at each run :frowning: we are totally baffled by this.

here is the example code which is executed:

```
  module_af = torch::jit::load(af_model_path);

    std::vector<torch::jit::IValue> inputs;
    at::Tensor input = torch::ones({1, 3, 112, 112});
    inputs.push_back(input);
    at::Tensor embeds0 = module_af.forward(inputs).toTensor();
```

The bizzare thing is, on a server we have and a small virtual machine on OVH, we do not see this problem - everything is consistent.

However when we try it on consumer grade laptops (2 machines to be precise), we get different results on each run.

here are some of the results we see:

```
Run 1, I get (just the head of the embeds0):

-0.132933
0.00335889
-0.283859
-0.127157
-0.0562173
-0.0433259
-0.0634281
-0.0114401
-0.0847342
-0.0239286
Run 2:

-0.194605
0.00335889
-0.283859
-0.127157
-0.0562173
-0.0433259
-0.0634281
-0.0114401
-0.0847342
-0.0239286
Run 3:

-0.194605
0.00335889
-0.283859
-0.127157
-0.0320272
-0.029183
-0.0418263
-0.0157832
-0.0434114
-0.015427
```

The jit traced model is here:

https://drive.google.com/file/d/1AbUPU9awdcljYZigXJbYjoA8Hh2NiMfx/view?usp=sharing

the model accepts a 1,3,112,112 input and produces a 1,512 output.

The original mode, from which the trace was produced using jit.trace is here:

https://drive.google.com/file/d/1iz9rIXMtZDQTgl95BEshr7WIzAR7li12/view?usp=sharing

We have tried using the pytorch forums to look for a solution, but no luck :(

Any pointers or tips on what to look for would be useful.

Thank you.

J

### Versions

python 1.6 and pytoch 1.7 (libtorch compiled from master)

cc @jbschlosser

**Comment by jbschlosser** [comment]
Hey @jamesrobertwilliams, there are several expected sources of non-determinism that could result in different outputs for the same input: use of dropout, CUDA-based non-determinism, etc. See the notes [here](https://pytorch.org/docs/stable/notes/randomness.html) for more details on reproducibility within PyTorch.

In general, the issues here are for problems within PyTorch itself, so we kindly suggest that you keep usage discussion like this to the forums. If you identify a potential issue within PyTorch itself, feel free to open an issue for that. Thank you!

**Comment by jamesrobertwilliams** [comment]
@mrshenli 
@jbschlosser with all due respect, you are totally wrong here. Please consider the following:

- the model is jit traced, and run on cpu.
- the model gives exactly the same results on consecutive runs on different (production type server) machines.
- the model gives different results on consecutive runs on consumer grade (laptops) hardware. The results as I show are wildly different.

I am very well aware of the issues you have pointed out such as drop out but this is not the issue here. The model has been jit traced.

**Comment by jbschlosser** [comment]
@jamesrobertwilliams I recommend the following:
1. Ensure you've placed the model in `eval()` mode before tracing to avoid e.g. updating batch norm stats. As I'm sure you're aware, the mode the model is placed in before tracing is the mode that the traced model will be in.
2. Run with the latest nightly version of PyTorch to see if your issue persists. If it does, please run the environment gathering script linked when you opened the issue so we have details on the environment and hardware in use.
3. It'd be helpful if you could investigate deeper into the specific modules causing the differing results. While I understand this can be difficult, ideally please provide a short, self-contained script that reproduces the issue you're seeing on the laptops.

**Comment by jamesrobertwilliams** [comment]
@jbschlosser thank you for reconsidering my issue :)

[1] The model is def in eval mode - as evidenced by my runs on other machines (where I get reproducible results across different servers - the results are EXACTLY the same on consecutive runs).,
[2] This was built directly from the master source. Please note that this is a libtorch issue (not pytorch) and thus these are my environments:

```
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 10 (buster) (x86_64)
GCC version: (Debian 8.3.0-6) 8.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.28

Python version: 3.7.3 (default, Jan 22 2021, 20:04:44)  [GCC 8.3.0] (64-bit runtime)
Python platform: Linux-4.19.0-20-amd64-x86_64-with-debian-10.12
Is CUDA available: N/A
CUDA runtime version: 11.2.152
GPU models and configuration: GPU 0: GeForce GTX 1060 with Max-Q Design
Nvidia driver version: 460.73.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.2
[conda] Could not collect
```

[3] 
Step:1: download jittraced model from here: 
https://drive.google.com/file/d/1AbUPU9awdcljYZigXJbYjoA8Hh2NiMfx/view?usp=sharing
Step:2: 

```
  torch::jit::script::Module module_af;
  module_af = torch::jit::load(af_model_path); // load model from path

    std::vector<torch::jit::IValue> inputs;
    at::Tensor input = torch::ones({1, 3, 112, 112});
    inputs.push_back(input);
    at::Tensor embeds0 = module_af.forward(inputs).toTensor();

    std::fstream res(embeds_logfile_name, std::ios::out | std::ios::binary); // write to file our cout
    for (int64_t i = 0; i < embeds0.numel(); ++i)
    {
      float a = embeds0[0][i].item<float>();
      res << a << std::endl;
    }
    res.close();
```

My cmake file:

```
# Minimal CMakeLists.txt for the AWS SDK for C++.
cmake_minimum_required(VERSION 3.8)
set(CMAKE_CXX_STANDARD 17)

# Use shared libraries, which is the default for the AWS C++ SDK build.
option(BUILD_SHARED_LIBS "Build shared libraries" ON)

if(NOT SDK_BUILD_SHARED_LIBS)
    set(SDK_BUILD_SHARED_LIBS OFF)
endif()

# "facer" is just an example value.
project(facer)
include_directories(include)
add_subdirectory(include/spdlog)
add_subdirectory(include/fmt)

## bellow is not needed since everything sites within include  now
# include_directories(cppitertools)

# Locate the AWS SDK for C++ package.
# "s3" and "cloudtrail" are just example values.
#find_package(AWSSDK REQUIRED COMPONENTS s3 cloudtrail)
find_package(Torch REQUIRED)
find_package( OpenCV REQUIRED )
# find_package(spdlog)  #header only so not req
# find_package(fmt) #header only so not req
find_package(Threads REQUIRED)

set(Boost_USE_STATIC_LIBS OFF) 
set(Boost_USE_MULTITHREADED ON)  
set(Boost_USE_STATIC_RUNTIME OFF) 
find_package(Boost COMPONENTS filesystem regex) 

add_executable(facer facer.cpp)

# Build using the C++ standard version 11.
target_compile_features(facer PUBLIC cxx_std_17)

# The libraries used by your executable.
# target_link_libraries(facer "${AWSSDK_LINK_LIBRARIES}")
target_link_libraries(facer "${TORCH_LIBRARIES}")
target_link_libraries (facer ${OpenCV_LIBS})
target_link_libraries(facer Boost::filesystem Boost::regex)
# target_link_libraries(facer spdlog) # see header only variant below
target_link_libraries(facer fmt)
target_link_libraries(facer spdlog::spdlog_header_only)

set_property(TARGET facer PROPERTY CXX_STANDARD 17)
```

Please let me know if I can provide anything else.

**Comment by jbschlosser** [comment]
Looks like your model consists of batch norm, conv, dropout, linear and prelu.. also confirmed that the serialized model correctly puts the batch norm and dropout layers into eval mode. So it seems like `Conv2d` is likely the culprit for non-determinism. @jamesrobertwilliams Perhaps you could dig into the backend selected across the two types of machines:
https://github.com/pytorch/pytorch/blob/4bf5380ec7fcf770b7a01c5d7fd4384a10a75510/aten/src/ATen/native/Convolution.cpp#L1289

**Comment by jamesrobertwilliams** [comment]
@jbschlosser firstly thank you ever so much for looking into it - so, I was not really insane :D
Now, in libtorch, how do I check the backend selected? is this something I specify when I compile libtorch?
Is it possible to print out what backend libtorch is using? Apologies if these are dumb questions :(

I am looking at this:
https://github.com/pytorch/pytorch/issues/18776
and this:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Context.cpp

and it looks like I can do `at::globalContext().setUserEnabledMkldnn(true);`
Am I in the right direction with this?

Once again, thank you ever so much for your help - I was losing my mind because of this issue.

**Comment by jbschlosser** [comment]
> PS: Were you able to reproduce the issue btw?

@jamesrobertwilliams I was not able to reproduce the issue - my results are consistent across runs. My current hypothesis is that your issue is related to the backend being selected on your hardware.

> Now, in libtorch, how do I check the backend selected? is this something I specify when I compile libtorch? Is it possible to print out what backend libtorch is using?

The backend is selected dynamically as a function of the input parameters. You'll have to make some changes to the source and recompile to print out the backend. The [diff here](https://github.com/pytorch/pytorch/files/8566797/diff.txt) will get you what you need.

**Comment by jamesrobertwilliams** [comment]
> @jamesrobertwilliams I was not able to reproduce the issue - my results are consistent across runs. My current hypothesis is that your issue is related to the backend being selected on your hardware.
> 

That's disappointing for me! the bizzare thing is there is another jit traced model I use within libtorch and that is totally consistent across all machines.I am pretty sure that has conv2d.

> The backend is selected dynamically as a function of the input parameters. You'll have to make some changes to the source and recompile to print out the backend. The [diff here](https://github.com/pytorch/pytorch/files/8566797/diff.txt) will get you what you need.

Thanks a tonne - I will look how to do this.

**Comment by jbschlosser** [comment]
> That's disappointing for me! the bizzare thing is there is another jit traced model I use within libtorch and that is totally consistent across all machines.I am pretty sure that has conv2d.

As the backend is selected dynamically based on the input parameters, the selected backend can be different across conv2d invocations.

Another thing: could you try `torch.jit.load()`ing the model from Python and seeing if the issue is there as well? I'd like to get a sense for if it is libtorch-specific or not.

**Comment by jamesrobertwilliams** [comment]
@jbschlosser I followed what you have said and did this in python:

```
cpu_device=torch.device('cpu')

# load trained model
model=torch.jit.load(model_loc, map_location=cpu_device)

inp=torch.ones([3,112,112]).unsqueeze(0)

output = model(inp)

print(output.shape)

np.savetxt("run05_pfc.out", output.detach().numpy(), delimiter="\t")

sys.exit()
```

I did 5 runs and they are all EXACTLY the same.

So yea, if it definitely a libtorch issue :(

**Comment by jamesrobertwilliams** [comment]
@jbschlosser I also wanted to add that I had built the libtorch from source using edited CMakeLists.txt. I turned off a few things to create a minimal CPU build, and I am wondering now if this is causing the issue in some machines. My edited CMakeLists is here: https://gist.github.com/jamesrobertwilliams/10c41ecd5a3bf69bff55435d74a11b40

I will meanwhile build a version with mkl-dnn - maybe this will help?

**Comment by jamesrobertwilliams** [comment]
@jbschlosser Think I have found out the cause for it :( I took the cpu version from pytorch's wenbsite - and built against this. Now I see repeatable results :( I wonder what change in my CMakeLists.txt caused it :( I will dig deeper.

I think you can close the issue - I apologise for wasting your time. It was clearly a mistake in my custom build :(

**Comment by jbschlosser** [comment]
@jamesrobertwilliams No worries - glad you got it figured out!

**Comment by lminer** [comment]
@jamesrobertwilliams did you ever figure out what the problem was?

**Comment by kostrykin** [comment]
Could it have been something related to the BLAS library that you link against?

It took me many hours to figure out why PyTorch was producing inconsistent results on some machines (afaik all Intel machines) while working nicely on others (afaik AMD). After setting `MKL_NUM_THREADS` to 1, the inconsistency went away, so I blame that there is a multithreading bug in MKL that only appears on Intel CPUs.

So if anyone else encounters this issue in the future, it might be worth checking what BLAS you are linking against, and if it is MKL, try running with `MKL_NUM_THREADS` set to 1.
